### PR TITLE
Simplify tests and do not use `eval`

### PR DESCRIPTION
**Pull request by devmotion**
This PR is basically a slightly modified copy of https://github.com/TuringLang/Bijectors.jl/pull/116. Instead of FiniteDiff it uses FiniteDifferences (since DistributionsAD uses it already), and I added some additional transformations to ensure that parameters and samples are always in the correct domains.

This PR is based on https://github.com/TuringLang/DistributionsAD.jl/pull/87 since it seemed weird to keep the special non-randomized tests for `Dirichlet` that couldn't capture the bug in https://github.com/TuringLang/DistributionsAD.jl/issues/86.

### CODE REVIEW

**Comment by devmotion** [comment]
Zygote tests passed (after 30 minutes instead of 1h 48min on master :smiley:), which indicates that https://github.com/TuringLang/DistributionsAD.jl/pull/87 is fine (Zygote tests there are still running).

**Comment by torfjelde** [comment]
So this looks all good to me, but that's because I already saw the same thing earlier :sweat_smile: Should I just accept it or should maybe someone else also have a look, _just_ to be sure?

**Comment by devmotion** [comment]
I requested a review from @mohamed82008 as well since I assume he implemented the current test setup.

**Comment by yebai** [comment]
@mohamed82008

**Comment by yebai** [comment]
thanks, @devmotion -- nice improvements!

**Comment by mohamed82008** [comment]
Sorry guys I was moving houses and had internet issues. Had a quick scan and looks good though.

File: test/ad/utils.jl [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                result = logpdf.(dist, xtilde)
+            else
+                result = logpdf(dist, xtilde)
+            end
+            return sum(result)
+        end
+    end
+
+    if isempty(θ)
+        # In this case we can only test the gradient with respect to `x`
```
Comment (mohamed82008) [inline test/ad/utils.jl:109]: We should test wrt to `x` only even if theta is not empty.

Comment (mohamed82008) [reply test/ad/utils.jl:109]: And we should check if the support is continuous first.

Comment (devmotion) [reply test/ad/utils.jl:109]: The logic is just copied from the previous implementation. There we test wrt `x` if theta is empty and don't check if the support is continuous (see e.g. https://github.com/TuringLang/DistributionsAD.jl/blob/master/test/test_utils.jl#L126 - somehow I forgot to delete the file...). I guess it's OK to not check if the support is continuous, since otherwise we couldn't test any derivatives at all and the `DistSpec` would not be useful. Additionally, both in the old and new implementation we test wrt `x` if theta is not empty and the support is continuous.

Comment (mohamed82008) [reply test/ad/utils.jl:109]: Even if the old implementation didn't test wrt x only with a non-empty theta, we still should. This is useful for prior distributions in Turing.

Comment (devmotion) [reply test/ad/utils.jl:109]: Hmm I can't follow, why would we want to remove tests? As long as tests pass in a reasonable amount of time (which they do with the new implementation), it should always be better to perform more tests, shouldn't it?

Comment (mohamed82008) [reply test/ad/utils.jl:109]: Yes I am suggesting performing more tests :) Currently I don't think we test wrt to `x` and only `x` when theta is not empty.

Comment (devmotion) [reply test/ad/utils.jl:109]: Ah I see, somehow I didn't get it and assumed you would want to remove some of the existing tests :laughing:
